File: app/src/main/java/com/manimarank/spell4wiki/utils/PrefManager.java
package com.manimarank.spell4wiki.utils;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;

import com.manimarank.spell4wiki.R;
import com.manimarank.spell4wiki.activities.LoginActivity;
import com.manimarank.spell4wiki.auth.AccountUtils;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class PrefManager {
    // shared pref
    private static final String PREF_NAME = "spell4wiki_pref";
    private static final int PRIVATE_MODE = 0;
    // Kay values
    private static final String USERNAME = "username";
    private static final String IS_LOGGED_IN_USER = "is_logged_in_user";
    private static final String IS_ANONYMOUS_USER = "is_anonymous_user"; // Only wiktionary use
    private static final String IS_FIRST_TIME_LAUNCH = "is_first_time_launch";
    private static final String LANGUAGE_CODE_SPELL_4_WIKI = "language_code_spell_4_wiki";
    private static final String LANGUAGE_CODE_SPELL_4_WORD_LIST = "language_code_spell_4_word_list";
    private static final String LANGUAGE_CODE_SPELL_4_WORD = "language_code_spell_4_word";
    private static final String LANGUAGE_CODE_WIKTIONARY = "language_code_wiktionary";
    private static final String LANGUAGE_CODE_APP = "language_code_app";
    private static final String UPLOAD_AUDIO_LICENSE = "upload_audio_license";
    private static final String COMMON_CATEGORIES = "common_categories";
    private static final String CSRF_TOKEN = "csrf_token";
    private static final String COOKIE = "cookie";
    private SharedPreferences pref;
    private SharedPreferences.Editor editor;
    private Context mContext;


    public PrefManager(Context context) {
        this.mContext = context;
        pref = mContext.getSharedPreferences(PREF_NAME, PRIVATE_MODE);
        editor = pref.edit();
        editor.apply();
    }

    // Is Login
    public void setUserSession(String name) {
        editor.putString(USERNAME, name);
        editor.putBoolean(IS_LOGGED_IN_USER, true);
        editor.putBoolean(IS_ANONYMOUS_USER, false);
        editor.apply();
    }

    public String getName() {
        return pref.getString(USERNAME, null);
    }

    public Boolean isLoggedIn() {
        return !pref.getBoolean(IS_ANONYMOUS_USER, false) && pref.getBoolean(IS_LOGGED_IN_USER, false) && AccountUtils.isLoggedIn();
    }

    public boolean isFirstTimeLaunch() {
        return pref.getBoolean(IS_FIRST_TIME_LAUNCH, true);
    }

    // First time launch for App Intro
    public void setFirstTimeLaunch(boolean isFirstTime) {
        editor.putBoolean(IS_FIRST_TIME_LAUNCH, isFirstTime);
        editor.apply();
    }

    public String getCsrfToken() {
        return pref.getString(CSRF_TOKEN, null);
    }

    public void setCsrfToken(String editToken) {
        editor.putString(CSRF_TOKEN, editToken);
        editor.apply();
    }

    public Boolean getIsAnonymous() {
        return pref.getBoolean(IS_ANONYMOUS_USER, false);
    }

    public void setIsAnonymous(Boolean isAnonymous) {
        editor.putBoolean(IS_ANONYMOUS_USER, isAnonymous);
        editor.apply();
    }

    /**
     * Clear session details when click logout
     */
    public void logoutUser() {
        boolean isFirstTime = isFirstTimeLaunch();
        String appLanguage = getAppLanguage();
        editor.clear();
        setFirstTimeLaunch(isFirstTime);
        setAppLanguage(appLanguage);

        // Remove sync - authenticator account
        AccountUtils.removeAccount();

        if (mContext != null && mContext instanceof Activity) {
            Activity activity = (Activity) mContext;
            // After logout redirect user to Login Activity
            activity.finishAffinity();
            activity.startActivity(new Intent(mContext, LoginActivity.class));
        }
    }

    private Set<String> getCookies() {
        return pref.getStringSet(COOKIE, null);
    }

    public void setCookies(Set<String> cookieList) {
        Set<String> existingCookie = getCookies();
        if (existingCookie != null && existingCookie.size() > 0)
            cookieList.addAll(existingCookie);
        editor.putStringSet(COOKIE, cookieList);
        editor.apply();
    }

    public String getLanguageCodeSpell4Wiki() {
        return pref.getString(LANGUAGE_CODE_SPELL_4_WIKI, "ta");
    }

    public void setLanguageCodeSpell4Wiki(String languageCode) {
        editor.putString(LANGUAGE_CODE_SPELL_4_WIKI, languageCode);
        editor.apply();
    }

    public String getLanguageCodeSpell4WordList() {
        return pref.getString(LANGUAGE_CODE_SPELL_4_WORD_LIST, "ta");
    }

    public void setLanguageCodeSpell4WordList(String languageCode) {
        editor.putString(LANGUAGE_CODE_SPELL_4_WORD_LIST, languageCode);
        editor.apply();
    }

    public String getLanguageCodeSpell4Word() {
        return pref.getString(LANGUAGE_CODE_SPELL_4_WORD, "ta");
    }

    public void setLanguageCodeSpell4Word(String languageCode) {
        editor.putString(LANGUAGE_CODE_SPELL_4_WORD, languageCode);
        editor.apply();
    }

    public String getLanguageCodeWiktionary() {
        return pref.getString(LANGUAGE_CODE_WIKTIONARY, "ta");
    }

    public void setLanguageCodeWiktionary(String languageCode) {
        editor.putString(LANGUAGE_CODE_WIKTIONARY, languageCode);
        editor.apply();
    }

    public String getUploadAudioLicense() {
        return pref.getString(UPLOAD_AUDIO_LICENSE, WikiLicense.LicensePrefs.CC_0);
    }

    public void setUploadAudioLicense(String uploadAudioLicense) {
        editor.putString(UPLOAD_AUDIO_LICENSE, uploadAudioLicense);
        editor.apply();
    }

    void setCommonCategories(List<String> categoryCommon) {
        editor.putStringSet(COMMON_CATEGORIES, new HashSet<>(categoryCommon));
        editor.apply();
    }

    public void setAppLanguage(String language){
        editor.putString(LANGUAGE_CODE_APP, language);
        editor.apply();
    }

    public String getAppLanguage() {
        return pref.getString(LANGUAGE_CODE_APP, "en");
    }


    public Set<String> getCommonCategories(){
        return pref.getStringSet(COMMON_CATEGORIES, null);
    }
}

File: app/src/main/java/com/manimarank/spell4wiki/data/prefs/PrefManager.kt
package com.manimarank.spell4wiki.data.prefs

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import com.manimarank.spell4wiki.data.auth.AccountUtils
import com.manimarank.spell4wiki.ui.login.LoginActivity
import com.manimarank.spell4wiki.utils.WikiLicense
import com.manimarank.spell4wiki.utils.constants.AppConstants

/**
 * Preference utility class for App level settings and login related preference
 */
class PrefManager(private val mContext: Context?) {
    private val pref: SharedPreferences
    private val editor: SharedPreferences.Editor

    /**
     * Set user logged in flag
     * @param name String?
     */
    fun setUserSession(name: String?) {
        editor.putString(USERNAME, name)
        editor.putBoolean(IS_LOGGED_IN_USER, true)
        editor.putBoolean(IS_ANONYMOUS_USER, false)
        editor.apply()
    }

    val name: String?
        get() = pref.getString(USERNAME, null)
    val isLoggedIn: Boolean
        get() = !pref.getBoolean(IS_ANONYMOUS_USER, false) && pref.getBoolean(IS_LOGGED_IN_USER, false) && AccountUtils.isLoggedIn

    // First time launch for App Intro
    var isFirstTimeLaunch: Boolean
        get() = pref.getBoolean(IS_FIRST_TIME_LAUNCH, true)
        set(isFirstTime) {
            editor.putBoolean(IS_FIRST_TIME_LAUNCH, isFirstTime)
            editor.apply()
        }
    var csrfToken: String?
        get() = pref.getString(CSRF_TOKEN, null)
        set(editToken) {
            editor.putString(CSRF_TOKEN, editToken)
            editor.apply()
        }
    var isAnonymous: Boolean?
        get() = pref.getBoolean(IS_ANONYMOUS_USER, false)
        set(isAnonymous) {
            editor.putBoolean(IS_ANONYMOUS_USER, isAnonymous!!)
            editor.apply()
        }

    /**
     * Clear session details when click logout
     */
    fun logoutUser() {
        clearLoginData()
        if (mContext != null && mContext is Activity) {
            val activity = mContext
            // After logout redirect user to Login Activity
            activity.finishAffinity()
            activity.startActivity(Intent(mContext, LoginActivity::class.java))
        }
    }

    /**
     * Clear login information once logged out
     */
    fun clearLoginData() {
        val isFirstTime = isFirstTimeLaunch
        editor.clear()
        isFirstTimeLaunch = isFirstTime

        // Remove sync - authenticator account
        AccountUtils.removeAccount()
    }

    private val cookies: Set<String>?
        get() = pref.getStringSet(COOKIE, null)

    /**
     * Set cookie information
     * @param cookieList MutableSet<String?>
     */
    fun setCookies(cookieList: MutableSet<String?>) {
        val existingCookie = cookies
        if (existingCookie != null && existingCookie.size > 0) cookieList.addAll(existingCookie)
        editor.putStringSet(COOKIE, cookieList)
        editor.apply()
    }

    var languageCodeSpell4WikiAll: String?
        get() = pref.getString(SPELL_4_WIKI_ALL, "ta")
        set(languageCode) {
            editor.putString(SPELL_4_WIKI_ALL, languageCode)
            editor.apply()
        }

    var uploadAudioLicense: String?
        get() = pref.getString(UPLOAD_AUDIO_LICENSE, WikiLicense.LicensePrefs.CC_0)
        set(uploadAudioLicense) {
            editor.putString(UPLOAD_AUDIO_LICENSE, uploadAudioLicense)
            editor.apply()
        }

    var runFilterNumberOfWordsToCheck: Int?
        get() = pref.getInt(RUN_FILTER_NO_OF_WORDS_TO_CHECK, AppConstants.RUN_FILTER_NO_OF_WORDS_CHECK_COUNT)
        set(noOfWordCount) {
            editor.putInt(RUN_FILTER_NO_OF_WORDS_TO_CHECK, noOfWordCount ?: AppConstants.RUN_FILTER_NO_OF_WORDS_CHECK_COUNT)
            editor.apply()
        }

    var isWiktionaryCleanupEnabled: Boolean
        get() = pref.getBoolean(WIKTIONARY_CLEANUP_ENABLED, true)
        set(isEnabled) {
            editor.putBoolean(WIKTIONARY_CLEANUP_ENABLED, isEnabled)
            editor.apply()
        }

    fun getWordsCategoryList(langCode: String?): MutableList<String> {
        return (pref.getStringSet(String.format(WORDS_CATEGORY_FOR_LANG, langCode ?: "ta"), null) ?: mutableSetOf()).toMutableList()
    }

    fun setWordsCategoryList(langCode: String?, categoryList: MutableSet<String>?) {
        editor.putStringSet(String.format(WORDS_CATEGORY_FOR_LANG, langCode ?: "ta"), categoryList)
        editor.apply()
    }

    fun getSelectedWordsCategory(langCode: String?): String? {
        return pref.getString(String.format(SELECTED_WORDS_CATEGORY_FOR_LANG, langCode ?: "ta"), null)
    }

    fun setSelectedWordsCategory(langCode: String?, category: String?) {
        editor.putString(String.format(SELECTED_WORDS_CATEGORY_FOR_LANG, langCode ?: "ta"), category)
        editor.apply()
    }

    companion object {
        // shared pref
        private const val PREF_NAME = "spell4wiki_pref"

        // Key values
        private const val USERNAME = "username"
        private const val IS_LOGGED_IN_USER = "is_logged_in_user"
        private const val IS_ANONYMOUS_USER = "is_anonymous_user" // Only wiktionary use
        private const val IS_FIRST_TIME_LAUNCH = "is_first_time_launch"
        private const val SPELL_4_WIKI_ALL = "language_code_spell_4_wiki_all"
        private const val UPLOAD_AUDIO_LICENSE = "upload_audio_license"
        private const val RUN_FILTER_NO_OF_WORDS_TO_CHECK = "run_filter_no_of_words_to_check"
        private const val WORDS_CATEGORY_FOR_LANG = "words_category_for_%s"
        private const val SELECTED_WORDS_CATEGORY_FOR_LANG = "selected_words_category_for_%s"
        private const val CSRF_TOKEN = "csrf_token"
        private const val COOKIE = "cookie"
        private const val WIKTIONARY_CLEANUP_ENABLED = "wiktionary_cleanup_enabled"
    }

    init {
        pref = mContext!!.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
        editor = pref.edit()
        editor.apply()
    }
}